### DISCUSSION TITLE
Possible to use engine from python and idea for app

### DISCUSSION THREAD

**Discussion by dstengle** [Tracktion/tracktion_engine#213] [Q&A]
I have an idea for an app, mostly headless that I'd like to write but haven't used c/c++ in decades. I'm wondering if it would be possible to use tracktion engine from python through something like popsicle or using pybind11. I'll be giving the c++ code a try with a lot of LLM assistance, but python would certainly make my life easier.

The app: a sampling clip grid looper and midi router. My goal is to be able to use a single synth like a microfreak with a hapax and all the tracks I could want with a simple interface. The hapax would use patterns with program changes to indicate which clip is active and a hardware interface (button per track or button/encoder) would tell the app which tracks are playing clips vs getting routed live midi. Transport would launch clips and record an armed track.

I don't necessarily need the new clip launching functionality. I just need midi handling, recording and playing in sync, which it seems engine should be able to do just fine.

I'm assuming this should all run fine on a raspberry pi and I have a couple of pisound cards or I could just start with usb. Prototyping would just be on a laptop to simplify things.

Nice additions would be autosampling for kits and instruments as well as effects,

If anyone has thoughts on other ways to achieve this I'd love to hear it. I looked at dawdreamer but there aren't any examples of live audio/midi input. Popsicle examples seemed to be all focused around graphics and I'm not sure how it would integrate but maybe it's possible to integrate engine as a module in popsicle.

**Comment by dstengle** [comment]
Looking at the apis it seems like a lot of the midi processing comes from JUCE and I could start start trying out at least some of the midi routing and see where I might need functionality from engine.

**Comment by julianstorer** [comment]
I'm currently working on a full javascript front-end for the engine, that's aimed at writing HTML/JS apps that sit on top of the engine. It's a work-in-progress, won't be available for a few months, but could be helpful.

But to be honest, interacting with a DAW engine requires a surprisingly large amount of API calls. Writing and maintaining the huge amount of boilerplate involved in translating all these quite complicated functions from one language to another could far outweigh any advantage you might get from using python.

**Reply by botoxparty** [reply]
Hey Julian,

I'd be very interested to contribute to the javascript front-end for the engine. I have a DAW that i originally started in plain JavaScript but I eventually just re-wrote the whole thing using JUCE and Tracktion due to performance limitations I was hitting.

What's the current state of the initiative? Do you have a repo I could look at?

**Reply by julianstorer** [reply]
It's very much a work in progress, but I'll probably make it public soon for people to get involved with..

**Reply by botoxparty** [reply]
Definitely ping me when you have something available, or if you want to share it privately I would love to have a look at the progress.
